ws/DiscordWebsocket.ts

```diff
@@ -0,0 +1,229 @@
+import WebSocket from 'ws';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { EVENTS, Payload, GATEWAY_OPCODES, GATEWAY_CLOSE_EVENT_CODES } from './constants';
+const IDENTIFY_TIMEOUT = 1000;
+
+interface ConnectionProperties {
+  $browser: string;
+  $device: string;
+  $os: string;
+}
+
+// TODO implement rate limits
+export default class DiscordWebsocket {
+  autoReconnect = true;
+  connectionProperties: ConnectionProperties;
+  discordTrace?: string;
+  largeThreshold?: number;
+  ready = false;
+  sessionID: string | null = null;
+  ws: WebSocket | null = null;
+  private _heartBeatInterval: NodeJS.Timeout | null = null;
+  private _intents: number;
+  private _lastHeartbeatAck = true;
+  private _selfDisconnect = false;
+  private _seq: number | null = null;
+  private _token: string;
+  private _url: string;
+
+  constructor(token: string, url: string, intents: number, connProps: Partial<ConnectionProperties> = {}) {
+    this._token = token;
+    this._url = url;
+    this._intents = intents;
+
+    this._identify = this._identify.bind(this);
+    this._onClose = this._onClose.bind(this);
+    this._onMessage = this._onMessage.bind(this);
+    this.connectionProperties = { $os: process.platform, $browser: 'dis.ts', $device: 'dis.ts', ...connProps };
+  }
+
+  connect() {
+    this.initialize();
+    return this;
+  }
+
+  disconnect(code?: number) {
+    if (code === 1000 || code === 1001) {
+      this.sessionID = null;
+      this._seq = null;
+    }
+    this._selfDisconnect = true;
+    this.ws?.close(code);
+    return this;
+  }
+
+  initialize() {
+    if (!this._url) throw new Error('Websocket URL not provided');
+
+    this._selfDisconnect = false;
+    if (this.ws) throw new Error('Attempted duplicate connection - run uninitialize');
+
+    this.ws = new WebSocket(this._url);
+    this.ws.once('close', this._onClose);
+    this.ws.on('message', this._onMessage);
+    this.ws.on('error', console.error);
+    return this;
+  }
+
+  reset() {
+    if (this._heartBeatInterval) clearInterval(this._heartBeatInterval);
+    if (this.ws?.readyState === WebSocket.OPEN) {
+      this.ws.off('message', this._onMessage)
+        .off('error', console.error)
+        .terminate();
+    }
+    this._heartBeatInterval = null;
+    this.ws = null;
+    this.ready = false;
+    this._lastHeartbeatAck = true;
+    return this;
+  }
+
+  restart(newSession = false) {
+    if (newSession) this.sessionID = null;
+    this.disconnect().connect();
+  }
+
+  // TODO Prevent 4012, 4013?
```

Comment:
for 4012, maybe an hardcoded fallback? (to v8 rn). This is only useful because we allow users to specificy their version themselves. It should display an error anyway. If fallback doesn't work, we should entire stop the gw connection and errors out. Maybe we should also error out anyway in the first case but I guess defaulting first is better.

for 4013 same behaviour? default to no intents or like guildMessage only. Otherwise error out entirely.

Honestly for both case we would be ok for erroring out entirely. Considering this is the low level wrapper it might make sense to just do that. On our higher level wrapper, we could do intents checking to make sure it's correct and defaulting and so on.

Reply:
> For 4013 same behaviour? default to no intents or like guildMessage only. Otherwise error out entirely.
Defaulting intents on a 4013 will cause a lot of WTFs for people not expecting, so not a fan. I'm happy to just error this out. 4012 makes more sense to fallback and I have implemented this but we can discuss implications and see whether it is worth falling back considering the nature of the subrepo